server/commands/list_objects.go

```diff
@@ -0,0 +1,234 @@
+	resolveNodeLimit             uint32
+}
+
+// NewListObjectsQuery creates a ListObjectsQuery
+func NewListObjectsQuery(datastore storage.OpenFGADatastore, tracer trace.Tracer, logger logger.Logger, meter metric.Meter, listObjectsDeadlineInSeconds int, listObjectsMaxResults uint32, ResolveNodeLimit uint32) *ListObjectsQuery {
+	return &ListObjectsQuery{
+		datastore:                    datastore,
+		logger:                       logger,
+		tracer:                       tracer,
+		meter:                        meter,
+		listObjectsDeadlineInSeconds: listObjectsDeadlineInSeconds,
+		listObjectsMaxResults:        listObjectsMaxResults,
+		resolveNodeLimit:             ResolveNodeLimit,
+	}
+}
+
+// Execute the ListObjectsQuery, returning a list of object IDs
+func (q *ListObjectsQuery) Execute(ctx context.Context, req *openfgapb.ListObjectsRequest) (*openfgapb.ListObjectsResponse, error) {
+	err := q.validateInput(ctx, req.StoreId, req.Type, req.AuthorizationModelId, req.Relation)
+	if err != nil {
+		return nil, err
+	}
+
+	resultsChan := make(chan string, q.listObjectsMaxResults)
+	errChan := make(chan error)
+	resolvedChan := make(chan struct{})
+
+	timeoutCtx := ctx
+	if q.listObjectsDeadlineInSeconds != 0 {
+		var cancel context.CancelFunc
+		timeoutCtx, cancel = context.WithTimeout(ctx, time.Second*time.Duration(q.listObjectsDeadlineInSeconds))
+		defer cancel()
+	}
+
+	go func() {
+		q.performChecks(timeoutCtx, req.StoreId, req.AuthorizationModelId, req.Type, req.Relation, req.User, req.ContextualTuples, resultsChan, errChan, resolvedChan)
+	}()
+
+	select {
+	case <-timeoutCtx.Done():
+		return nil, serverErrors.NewInternalError("Timeout exceeded", timeoutCtx.Err())
+	case <-resolvedChan:
+		objectIDs := make([]string, 0)
+		for objectID := range resultsChan {
+			objectIDs = append(objectIDs, objectID)
+		}
+
+		return &openfgapb.ListObjectsResponse{
+			ObjectIds: objectIDs,
+		}, nil
+	case genericError := <-errChan:
+		return nil, genericError
+	}
+}
+
+func (q *ListObjectsQuery) ExecuteStreamed(ctx context.Context, req *openfgapb.StreamedListObjectsRequest, srv openfgapb.OpenFGAService_StreamedListObjectsServer) error {
+	err := q.validateInput(ctx, req.StoreId, req.Type, req.AuthorizationModelId, req.Relation)
+	if err != nil {
+		return err
+	}
+
+	resultsChan := make(chan string, q.listObjectsMaxResults)
+	errChan := make(chan error)
+	resolvedChan := make(chan struct{})
+
+	timeoutCtx := ctx
+	if q.listObjectsDeadlineInSeconds != 0 {
+		var cancel context.CancelFunc
+		timeoutCtx, cancel = context.WithTimeout(ctx, time.Second*time.Duration(q.listObjectsDeadlineInSeconds))
+		defer cancel()
+	}
+
+	go func() {
+		q.performChecks(timeoutCtx, req.StoreId, req.AuthorizationModelId, req.Type, req.Relation, req.User, req.ContextualTuples, resultsChan, errChan, resolvedChan)
+	}()
+
+	for {
+		select {
+		case <-timeoutCtx.Done():
+			return serverErrors.NewInternalError("Timeout exceeded", timeoutCtx.Err())
```

Comment:
We shouldn't return an error here. If the timeout was exceeded then we're done. The timeout deadline is merely to control the maximum amount of time to spent streaming/aggregating the results. If we've hit it, then we just return what we've accumulated up to that point.

Reply:
Ah you're right it doesn't make sense for the streaming version. Removed it.

For the regular API however, if I insert an artificial delay of 5 seconds in one of the Checks, it doesn't matter what I put here: the whole request aborts at the 3 second mark with "context timeout exceeded" which I think is coming from https://github.com/openfga/openfga/blob/65c5f3bd3d50eb22b2a7fc0f41051eba9631da2e/pkg/cmd/service/service.go#L172

Reply:
It is, yeah. We should probably default the UpstreamTimeout to infinity instead of 3sec. I think if `UpstreamTimeout` is more strict that the `ListObjectsDeadline` than this behavior is expected. But I think the default should be to not have an UpstreamTimeout unless it's specified.

Reply:
👍 returning partial results now.